### ISSUE TITLE
Consider updating SQLite min (and default) supported version

### ISSUE THREAD

**Issue by thomcc** [jgallagher/rusqlite#706]
Right now it's 3.6.8. This is from 2009! We also offer 4 different pregenerated versions of headers up through 3.7.7... which is from 2011.

We should pick a version that makes our lives easier to support. @gwenn, do you have any thoughts or requirements here? Anybody else?

Worth noting that we could do away with the headers, and always use one which is up to date, and avoid some build complexity. Rusqlite would need to avoid calling functions that aren't present, but it already does that. If you don't call an extern "C" function, it isn't linked. SQLite is extremely ABI stable, so this is safe. `bindgen` is just paranoid about libraries less well-behaved.

**Comment by gwenn** [comment]
https://github.com/jgallagher/rusqlite/issues/293#issuecomment-326995795

**Comment by thomcc** [comment]
In https://github.com/launchbadge/sqlx/issues/191 it's been noted that support back to v3.8.11 is desirable.

In that issue I dumped some of my thoughts on this subject which I'll reproduce here for ease of referencing.

> I've meant to reach out to you about this, and do a write up, but we have an open issue in the rusqlite repository around the minimum/default supported version: https://github.com/rusqlite/rusqlite/issues/706. (FWIW there's no rule that says that rusqlite and libsqlite3-sys need to have the same values here, although it would certainly be convenient for me if they did).
> 
> That said it's been a rough year in general and I haven't found the time to do so yet. That said, if you have opinions here they'd be valuable. Even the knowledge that you care about 3.8.11+ is very helpful.
> 
> In general I think the way libsqlite3-sys / rusqlite handles versions to be suboptimal. In an ideal world, there would be only a few versions that control underlying features, and then we'd have feature flags for them. `clang-sys` does this like https://github.com/KyleMayes/clang-sys/blob/master/Cargo.toml#L22-L42 and is fairly okay as a result, although you can see the fractures of that approach starting to not scale very well.
> 
> However, the world we live in is one in which there are very very many versions we would have to do this way, far more than clang-sys has to deal with...

Anyway, diesel is another quite large user of libsqlite3-sys in the ecosystem, so: @sgrif / @weiznich do you have any thoughts on this topic? Even "I'd prefer no change to the status quo for libsqlite3-sys" is useful feedback. No rush at all here, just trying to take stock of the requirements/options here.

**Comment by mehcode** [comment]
> In launchbadge/sqlx#191 it's been noted that support back to v3.8.11 is desirable.

I just want to clarify that I meant "at least 3.8.11". I have no preference when it comes to "support of older versions". If I was making the decision, I would follow whatever stance SQlite has on security fixes.

**Comment by thomcc** [comment]
Ah, that makes sense, and is a good idea. I'll look into that, thanks.

**Comment by thomcc** [comment]
> I would follow whatever stance SQlite has on security fixes

I've asked around and SQLite's stance here is that they don't publish security fixes for old releases but new releases are always backwards compatible, so updating is easy.

I think I'd like to be slightly more pragmatic about this (since we're in a slightly different situation being bindings), unfortunately, but will give it more thought.

**Comment by weiznich** [comment]
Diesel does compile `libsqlite3-sys` with the `min_sqlite_version_3_7_16` feature enabled. As far as I've checked the functions we are using that (3.7.16) should be the minimal supported sqlite3 version from a ABI stand point. Additionally there is the possibility to use some SQL functionality that requires newer sqlite version, but if you don't use that you should be fine in using diesel with sqlite version 3.7.16. That written: I'm not sure if someone actually uses diesel which such an old version and we do not run tests against that version, so it could be broken. I think I would be fine to just have generated bindings to an up to date version, as long as it is possible to link older versions if you only use functions that are supported on that old version.

**Comment by thomcc** [comment]
Yeah, we should just move to providing a up to date binding definitions, since you can still link to older versions of SQLite with these bindings, so long as you only actually *call* functions available in that version (otherwise you'll get linker errors).

I wasn't around then, but I believe part of the original idea was this way users of libsqlite3-sys need to "opt-in" to use newer features by accident, but IMHO this causes more complexity than it saves. (That, or it's possible they thought thought having the functions defined would cause the linker failure even if unused, who knows).

After this change, it would be up to users of libsqlite3-sys to ensure they don't use functions and APIs from newer versions of SQLite than they want (I usually check [sqlite3ext.h](https://github.com/rusqlite/rusqlite/blob/840a6f3789f91d71599fa8fe909cc0d51b13f862/libsqlite3-sys/sqlite3/sqlite3ext.h#L625-L644) to figure this out).

This is probably fine, libsqlite3-sys is a very low level crate, and this helps avoid stagnation in the future. (I think we'd also give up the policy of "update to bundled bindings == breaking change for libsqlite3-sys", although that's probably not strictly related)

---

Concretely, this would look like libsqlite3-sys losing the features for:

- "min_sqlite_version_3_6_8"
- "min_sqlite_version_3_6_23"
- "min_sqlite_version_3_7_7"
- "min_sqlite_version_3_7_16"
- "bundled_bindings"

The last of which would become the new default behavior, which would be used unless overridden by a feature like "buildtime_bindgen", "sqlcipher", etc.

(Note: These changes basically only apply to libsqlite3-sys, and not rusqlite (with the exception of "bundled_bindings" which would become redundant in the same way). That is, rusqlite would probably still support ancient-as-hell sqlite for a bit longer, mostly since I still have no clue what the best version to target is.)

---

Anybody have objections to this? (@gwenn, @mehcode, @weiznich)

**Comment by weiznich** [comment]
First of all I believe removing this features would be a breaking change. If that's combined with a corresponding version change that's fine. For diesel specifically this would mean that we cannot update to any newer `libsqlite3-sys` version for the `1.x` release anymore as this again would be a breaking change. As we already plan that the next release would be a major release (i.e `2.0`) that would be fine for us as long as we somehow coordinate those releases a bit. (That written I cannot really comment on when the 2.0 release will exactly happen, other than hopefully this year)

Beside of that: Is there any reason why future releases could require a breaking change? If not, what's about cutting this change as 1.0 release of `libsqlite3-sys`? At least from diesels point of view there seams to be no reason why there are that many semver incompatible versions out there, as we already support a [really large version range](https://github.com/diesel-rs/diesel/blob/3e8fd4d7ce257830f1735e9a31c9a5f0fb0d777f/diesel/Cargo.toml#L19). Having not to increase the maximal version every time a new release is cut would be definitely welcome.

**Comment by thomcc** [comment]
> First of all I believe removing this features would be a breaking change. If that's combined with a corresponding version change that's fine.

Yes, we're in agreement.

> For diesel specifically this would mean that we cannot update to any newer libsqlite3-sys version for the 1.x release anymore as this again would be a breaking change

Hm, so you have libsqlite3-sys types in your public API? Tricky. We historically have bumped the major version of libsqlite3-sys when we updated the bundled version. I'm not *really* a fan of that, and I'm not exactly sure why it is, and that's what I meant by:

> I think we'd also give up the policy of "update to bundled bindings == breaking change for libsqlite3-sys", and only do breaking changes if the change corresponds to a breaking rust change — probably not strictly the same, but related

One thing worth noting is that breaking changes for C and Rust don't line up exactly right. SQLite considers it non-breaking to add new fields to structs and the like, and so even if we did do this as `1.0.0` of libsqlite3-sys, its possible we'd have to make further breaking changes in the future when SQLite updates in ways that are not fully compatible with Rust.

**Comment by weiznich** [comment]
> Hm, so you have libsqlite3-sys types in your public API? Tricky.

Diesel does not expose any `libsqlite3-sys` type via it's public API. Nerveless I consider changing this range of supported versions of this dependency as breaking change as cargo only allows one version of an crate that links to a native library. That means if some user depends on diesel with the sqlite feature enabled with the current version range (">=0.8.0, <0.23.0") and some version of rusqlite which uses a compatible `libsqlite3-sys` version internally compilation will work fine, as cargo picks exactly one version. As soon as we drop one of the supported versions that could break as there is now no single matching version anymore, so it's a breaking change for diesel to remove versions from that range. That combined with the fact that cargo will not pull in a `libsqlite3-sys` version that does not have the already used feature flag (see https://github.com/rusqlite/rusqlite/issues/454 for details) will make it impossible for diesel 1.x to ever support the new version.

>  One thing worth noting is that breaking changes for C and Rust don't line up exactly right. SQLite considers it non-breaking to add new fields to structs and the like, and so even if we did do this as 1.0.0 of libsqlite3-sys, its possible we'd have to make further breaking changes in the future when SQLite updates in ways that are not fully compatible with Rust.

That's not what I've hoped for, but you are right there is no way to handle that without breaking changes.
